### PR TITLE
MirrorGarden output may include Adapton-produced graphics

### PR DESCRIPTION
**Pull request by matthewhammer**
- Use motoko-adapton in the MirrorGarden repo, to ensure it can integrate into CI, `ic-mt`, etc.
- Demoable in two modes now, crudely --- Text editor demo when the window is wide, and Adapton demo with narrow :)
- (A next step is a real interactive menu system, and event capture/replay for CI, and scripts that explore each demo by exercising this menu system. For now things are crude.)
- Upgrade dfx and moc and base

### bug report 

**Note: now [found and fixed](dfinity/candid#155).**

This new demo crashes my use of `candid` with an error about an infinite value.

```
Candid decoding error: Deserialize("There is an infinite loop in the record definition, the type is isomorphic to an empty type", "Trailing type: [I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10)]\nTrailing value: [19, 01, 03, 03, 00, 00, 00, 00, 00, 00, 01, 03, 03, 00, 03, 00, 00, 00, 00, 01, 03, 03, 00, 06, 00, 00, 00, 00, 01, 03, 03, 00, 09, 00, 78, 78, 00, 01, 03, 03, 00, 0c, 00, 00, 00, 00, 01, 03, 03, 03, 00, 00, 00, 00, 00, 01, 03, 03, 03, 03, 00, 00, 00, 00, 01, 03, 03, 03, 06, 00
```

[The offending line in ic-mt](https://github.com/matthewhammer/ic-mini-terminal/blob/d07a33118c287ae40429e4349b0b60ad467fc18a/src/bin/ic-mt.rs#L800) uses a candid macro to decode the graphics, which seems to fail here.

Notably, `ic-mt` worked fine for other demos (e.g., the one modified here in this PR) and used the same type and similar-looking values (trees of rectangles).

It's not clear yet what is novel about this program, but the issue seems like it's within the Rust-based Candid deserializer logic, not these tools.  Still not sure though.

### CODE REVIEW

File: scripts/MirrorGarden-start-replica.sh [hunk 1]
```diff
@@ -9,8 +9,9 @@ dfx start --background --clean &&\
 dfx canister create MirrorGarden &&\
 dfx build MirrorGarden &&\
 dfx canister install MirrorGarden
-ic-gt connect 127.0.0.1:8000 `dfx canister id MirrorGarden` --user '("Alice", (100, 200, 200))' &
-ic-gt connect 127.0.0.1:8000 `dfx canister id MirrorGarden` --user '("Bob", (200, 100, 200))' &
+
+# ic-mt connect 127.0.0.1:8000 `dfx canister id MirrorGarden` --user '("Alice", (100, 200, 200))'
```
Comment (matthewhammer) [inline scripts/MirrorGarden-start-replica.sh]: These are the lines that crash now, with this message:

```
Candid decoding error: Deserialize("There is an infinite loop in the record definition, the type is isomorphic to an empty type", "Trailing type: [I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10), I(4), U(1136381571), I(7), U(1269255460), I(10)]\nTrailing value: [19, 01, 03, 03, 00, 00, 00, 00, 00, 00, 01, 03, 03, 00, 03, 00, 00, 00, 00, 01, 03, 03, 00, 06, 00, 00, 00, 00, 01, 03, 03, 00, 09, 00, 78, 78, 00, 01, 03, 03, 00, 0c, 00, 00, 00, 00, 01, 03, 03, 03, 00, 00, 00, 00, 00, 01, 03, 03, 03, 03, 00, 00, 00, 00, 01, 03, 03, 03, 06, 00
```

Comment (chenyan-dfinity) [reply scripts/MirrorGarden-start-replica.sh]: Can you rebuild candid with https://github.com/dfinity/candid/pull/155?

It adds the counter in the error message. If it's a concurrency bug, the counter would differ by a lot.

Comment (matthewhammer) [reply scripts/MirrorGarden-start-replica.sh]: Done.  It differs by one (17 versus 16).

Comment (matthewhammer) [reply scripts/MirrorGarden-start-replica.sh]: Notably, I also get the same numbers even when I run `ic-mt` with `-t`, which enables trace-level logging, and slows things down.  It doesn't seem likely that it's concurrency, based on that observation just now.

Comment (chenyan-dfinity) [reply scripts/MirrorGarden-start-replica.sh]: I pushed the fix in the candid branch, can you try again?

Comment (matthewhammer) [reply scripts/MirrorGarden-start-replica.sh]: The fix worked!  Thanks!

image


That's the graphics, as produced by `ic-mt`, on this PR now.

**Pull request merged**